bytecomp/symtable.ml

```diff
@@ -404,7 +404,43 @@ open Format
 
 let report_error ppf = function
   | Undefined_global s ->
-      fprintf ppf "Reference to undefined global `%s'" s
+      fprintf ppf "Reference to undefined global `%s'" s;
+      if !Sys.interactive then
+      begin
+        fprintf ppf
+          "@.Hint: @[\
+           This means a module's interface is loaded, \
+           but its implementation is not.@,";
+        begin match Load_path.find_uncap (s ^ ".cmo") with
+        | path ->
+          fprintf ppf
+            "Found %s @,in the load paths. \
+             @,Did you mean to load it with @,#load \"%s\" \
+             @,or by passing it as an argument to the toplevel?"
+             path (Filename.basename path)
+        | exception Not_found ->
+        match Load_path.find_uncap (s ^ ".cma") with
+        | path ->
+          fprintf ppf
+            "Found %s @,in the load paths. \
+             @,Did you mean to load it with @,#load \"%s\" \
+             @,or by passing it as an argument to the toplevel?"
+             path (Filename.basename path)
+        | exception Not_found ->
+        match Load_path.find_uncap (s ^ ".ml") with
+        | path ->
+          fprintf ppf
+            "Found %s @,in the load paths. \
+             @,Did you mean to compile it and load the produced bytecode file \
+             @,or to execute it with #mod_use \"%s\"?"
+             path (Filename.basename path)
+        | exception Not_found ->
+          fprintf ppf
+            "Did you mean to load a compiled implementation of the module \
+             @,with #load or by passing it as an argument to the toplevel?"
```

Comment:
One aspect of this new code that I don't like is the repetition of the messages in the `.cmo` and `.cma` cases. You could:
- Factor the two tests in one. This can be done for example by adding a `Load_path.find_uncap_opt` function and then
   ```ocaml
   List.find_map Load_path.find_uncap_opt [s ^ ".cmo"; s ^ ".cma"]
   ```
- Or abstract the printing logic as auxiliary functions first, and call the same function twice.

   ```ocaml
   let start_hint ppf = ... in
   let found_matching_object ppf path = ... in
   let found_matching_source ppf path = ... in
   let generic_advice ppf = ... in
   ```

Reply:
I chose to just define a local helper instead of adding a `Load_path.find_uncap_opt`, because I would probably also have to add a useless for now `Load_path.find_opt` for consistency.